### ISSUE TITLE
Support request: GF63 Thin 11UC but with a newer firmware version

### ISSUE THREAD

**Issue by iNUCi** [BeardOverflow/msi-ec#114]
### Laptop model

GF63 Thin 11UC (i5-11400H, RTX 3050 mobile, 16GB RAM)

### EC firmware version

16R6EMS1.107

### EC memory dump

```
     | _0 _1 _2 _3 _4 _5 _6 _7 _8 _9 _a _b _c _d _e _f
-----+------------------------------------------------
0x0_ | 00 80 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x1_ | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x2_ | 00 00 00 00 00 00 00 00 0a 05 00 00 00 04 09 0b
0x3_ | 03 09 00 0d 00 00 50 81 94 11 88 2c c2 01 e0 00
0x4_ | 00 00 62 00 b0 0e 00 00 48 0e c6 30 e7 0b 00 00
0x5_ | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0x6_ | 00 00 00 00 00 00 00 00 33 00 37 40 49 4c 52 58
0x7_ | 64 2b 26 2b 30 36 3c 46 55 64 08 03 03 03 03 03
0x8_ | 00 00 37 3d 43 49 4f 55 63 26 26 2b 30 36 3c 46
0x9_ | 55 64 08 03 03 03 03 03 06 0f 64 06 0f 64 41 00
0xa_ | 31 36 52 36 45 4d 53 31 2e 31 30 37 30 36 32 38
0xb_ | 32 30 32 33 31 38 3a 31 32 3a 31 38 00 00 00 28
0xc_ | 00 00 07 25 00 00 00 00 00 d3 00 00 00 00 00 00
0xd_ | 00 00 c1 83 0d 00 05 80 00 01 00 00 00 02 00 00
0xe_ | e2 00 00 89 0e 00 00 40 00 00 00 00 00 c0 00 00
0xf_ | 40 00 70 00 00 64 00 00 64 00 00 00 00 00 00 00
```

### GPU

Nvidia

### Is your keyboard RGB?

No (single color)

### Additional context

Currently, this module only seems to support an older version of this laptop's EC firmware and tells me that my firmware version is not supported

**Comment by glpnk** [comment]
Dumps looks same with older version. Do you use MSI-EC before? Do you update bios? If yes - can you send patch note from update?

For quick start, you can manually add your version to source or just change old version to new and install it. https://github.com/BeardOverflow/msi-ec/blob/6feed4617bafc9f6911f122395ab148e6e717389/msi-ec.c#L1000-L1003

**Comment by iNUCi** [comment]
Yes, i have updated the BIOS to E16R6IMS.113 from the MSI website
The file attached is the patch note that came with the BIOS update file
[E16R6IMS.1xx.txt](https://github.com/BeardOverflow/msi-ec/files/15058235/E16R6IMS.1xx.txt)

I think i remember trying this module on an older BIOS version and it successfully loading and working

**Comment by glpnk** [comment]
Thanks, changelog mention `106` and `107` versions, which could be same like `104`. I PR it later, or you can if you want

**Comment by teackot** [comment]
Neither .106 nor .107 seem to have breaking EC interface changes.

Merged in #117

You can now test the latest version! Also, can you please check if fn_win_swap works? (the values might also be reversed)

**Comment by Iampotato204** [comment]
uname: Linux debian 6.11.5+bpo-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.11.5-1~bpo12+1 (2024-11-11) x86_64 GNU/Linux
(debian bookworm + linux-image from bookworm-backports - i wanted to use msi-ec, but there was no folder /sys/devices/platform/msi-ec)
built from github using make, as dkms not enabled
battery_mode is unknown (see info below)

`==> ./available_fan_modes <==
auto
silent
advanced

==> ./available_shift_modes <==
eco
comfort
sport
turbo

==> ./battery_mode <==
unknown (128)

==> ./cooler_boost <==
off

==> ./driver_override <==
(null)

==> ./fan_mode <==
auto

==> ./fn_key <==
left

==> ./fw_release_date <==
2021/11/14 15:56:31

==> ./fw_version <==
16R6EMS1.104

==> ./modalias <==
platform:msi-ec

==> ./shift_mode <==
comfort

==> ./uevent <==
DRIVER=msi-ec
MODALIAS=platform:msi-ec

==> ./webcam <==
off

==> ./webcam_block <==
off

==> ./win_key <==
right
`

**Comment by Iampotato204** [comment]
> Neither .106 nor .107 seem to have breaking EC interface changes.
> 
> Merged in #117
> 
> You can now test the latest version! Also, can you please check if fn_win_swap works? (the values might also be reversed)

Yes, values ARE indeed reversed, setting 'win_key' to 'left' makes it work on right

**Comment by Iampotato204** [comment]
ok, and battery_mode is set to 'unknown (128)', why is that? Default value is written incorrectly or msi-ec handles it wrong?

**Comment by Iampotato204** [comment]
> yes, it's okay, some settings are persistent only because BIOS set them during power on

i suppose it wouldn`t be very wise to flash modified firmware... is there an acceptable workaround for this?

**Comment by glpnk** [comment]
You can't make custom firmware for closed source 8051 MCU without proprietary sources, toolchain and compiler

**Comment by Iampotato204** [comment]
> You can't make custom firmware for closed source 8051 MCU without proprietary sources, toolchain and compiler

yeah, so i mean, what is the best way to set it on pc boot?



> > ok, and battery_mode is set to 'unknown (128)', why is that? Default value is written incorrectly or msi-ec handles it wrong?
> 
> Probably never set - default 100%

it does indeed charge to 100

**Comment by glpnk** [comment]
> yeah, so i mean, what is the best way to set it on pc boot?

Some features set by bios on boot, some persist between reboot, but it's hard to tell

> it does indeed charge to 100

you can call it unset, for me max % persist between reboots

**Comment by glpnk** [comment]
Fixed FN-WIN inversion
